# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/mediasink/AudioSinkWrapper.h
# Commit: acb4ae1c1e57
# Full Hash: acb4ae1c1e57de7130cbfb499d6c30dcf134dfbb
# Author: Paul Adenot <paul@paul.cx>
# Date: 2022-05-24 21:44:48
# Regressor Bug: 1743834
# File Overlap Count: 1
# Description:
#   Bug 1743834 - Shut down the AudioSink when audio is muted. r=alwu,media-playback-reviewers
#   
#   This does the following:
#   - When the media is muted, shut down and release the AudioSink ;
#   - While the media is muted, use the system clock to make video advance ;
# ==============================================================================

diff -r 0b20ed64f57c -r acb4ae1c1e57 dom/media/mediasink/AudioSinkWrapper.h
--- a/dom/media/mediasink/AudioSinkWrapper.h	Tue May 24 13:09:05 2022 +0000
+++ b/dom/media/mediasink/AudioSinkWrapper.h	Tue May 24 13:09:06 2022 +0000
@@ -66,6 +66,7 @@
   media::TimeUnit GetPosition(TimeStamp* aTimeStamp = nullptr) override;
   bool HasUnplayedFrames(TrackType aType) const override;
   media::TimeUnit UnplayedDuration(TrackType aType) const override;
+  void DropAudioPacketsIfNeeded(const media::TimeUnit& aMediaPosition);
 
   void SetVolume(double aVolume) override;
   void SetStreamName(const nsAString& aStreamName) override;
@@ -86,6 +87,8 @@
   void GetDebugInfo(dom::MediaSinkDebugInfo& aInfo) override;
 
  private:
+  bool IsMuted() const;
+  void OnMuted(bool aMuted);
   virtual ~AudioSinkWrapper();
 
   void AssertOwnerThread() const {
@@ -94,7 +97,11 @@
 
   nsresult StartAudioSink(const media::TimeUnit& aStartTime);
 
-  media::TimeUnit GetVideoPosition(TimeStamp aNow) const;
+  // Get the current media position using the system clock. This is used when
+  // the audio is muted, or when the media has no audio track. Otherwise, the
+  // media's position is based on the clock of the AudioStream.
+  media::TimeUnit GetSystemClockPosition(TimeStamp aNow) const;
+  bool CheckIfEnded() const;
 
   void OnAudioEnded();
 
